Configuration cache error: Class 'org.gradle.internal.classpath.Instrumented$$Lambda$2709/0x00000098020e55e0' cannot be encoded

### Current Behavior

Configuration cache error: Build file 'build.gradle.kts': Class 'org.gradle.internal.classpath.Instrumented$$Lambda$2709/0x00000098020e55e0' cannot be encoded because class loader 'VisitableURLClassLoader(ant-and-gradle-loader)' of type 'org.gradle.internal.classloader.VisitableURLClassLoader' could not be encoded and the class is not available through the default class loader.

### Expected Behavior

No error.

### Context (optional)

_No response_

### Self-contained Reproducer Project

[project.zip](https://github.com/user-attachments/files/19011349/project.zip)

```kotlin
// build.gradle.kts

abstract class SystemGetEnvSource :
  ValueSource<Map<String, String?>, ValueSourceParameters.None> {

  override fun obtain(): Map<String, String?> = System.getenv()
}

val currentVars: Provider<Map<String, String?>> = providers.of(SystemGetEnvSource::class.java) { }

println("got ${currentVars.get().size} env vars")
```

Run `./gradlew help`

Observe configuration cache storage problem:

```
Parallel Configuration Cache is an incubating feature.
Calculating task graph as no cached configuration is available for tasks: help

> Configure project :
got 58 env vars

> Task :help

Welcome to Gradle 8.13.

To run a build, run gradlew <task> ...

To see a list of available tasks, run gradlew tasks

To see more detail about a task, run gradlew help --task <task>

To see a list of command-line options, run gradlew --help

For more detail on using Gradle, see https://docs.gradle.org/8.13/userguide/command_line_interface.html

For troubleshooting, visit https://help.gradle.org

[Incubating] Problems report is available at: file:///Users/dev/projects/tmp/gradle-classloader-error-weird/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Configuration cache problems found in this build.

1 problem was found storing the configuration cache.
- Build file 'build.gradle.kts': Class 'org.gradle.internal.classpath.Instrumented$$Lambda$2709/0x00000098020e55e0' cannot be encoded because class loader 'VisitableURLClassLoader(ant-and-gradle-loader)' of type 'org.gradle.internal.classloader.VisitableURLClassLoader' could not be encoded and the class is not available through the default class loader.
These are the known class loaders:
	- VisitableURLClassLoader(ClassLoaderScopeIdentifier.Id{coreAndPlugins:settings[:]:buildSrc[:]:root-project[:]:kotlin-dsl:/Users/dev/projects/tmp/gradle-classloader-error-weird/build.gradle.kts:Project/TopLevel/stage2(local)})
	- VisitableURLClassLoader(ClassLoaderScopeIdentifier.Id{coreAndPlugins:kotlin-dsl:/Users/dev/projects/tmp/gradle-classloader-error-weird/settings.gradle.kts:Settings/TopLevel/stage1(local)})
	- VisitableURLClassLoader(ClassLoaderScopeIdentifier.Id{coreAndPlugins:settings[:]:buildSrc[:]:kotlin-dsl:/Users/dev/projects/tmp/gradle-classloader-error-weird/build.gradle.kts:Project/TopLevel/stage1(local)})
Please report this error, run './gradlew --stop' and try again.

See the complete report at file:///Users/dev/projects/tmp/gradle-classloader-error-weird/build/reports/configuration-cache/bz7dzvjm2ewa3cqfxqbu4wptt/92foduwdx2d92cbjlly4vqi58/configuration-cache-report.html
> Class 'org.gradle.internal.classpath.Instrumented$$Lambda$2709/0x00000098020e55e0' not found in Gradle runtime class loader 'MixInLegacyTypesClassLoader(legacy-mixin-loader)' of type 'org.gradle.initialization.MixInLegacyTypesClassLoader'.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 331ms
1 actionable task: 1 executed
Configuration cache entry discarded with 1 problem.
```

### Gradle version

8.13

### Build scan URL (optional)

_No response_

### Your Environment (optional)

_No response_

--- a/.teamcity/subprojects.json
+++ b/.teamcity/subprojects.json
@@ -647,7 +647,7 @@
     "name": "input-tracking",
     "path": "platforms/core-configuration/input-tracking",
     "unitTests": true,
-    "functionalTests": false,
+    "functionalTests": true,
     "crossVersionTests": false
   },
   {

--- a/platforms/core-configuration/input-tracking/build.gradle.kts
+++ b/platforms/core-configuration/input-tracking/build.gradle.kts
@@ -7,4 +7,6 @@ description = "Configuration input discovery code"
 dependencies {
     api(libs.jspecify)
     api(libs.guava)
+
+    integTestDistributionRuntimeOnly(projects.distributionsCore)
 }

--- a/platforms/core-configuration/input-tracking/src/integTest/groovy/org/gradle/internal/configuration/inputs/InputTrackingIntegrationTest.groovy
+++ b/platforms/core-configuration/input-tracking/src/integTest/groovy/org/gradle/internal/configuration/inputs/InputTrackingIntegrationTest.groovy
@@ -1,0 +1,73 @@
+/*
+ * Copyright 2025 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.configuration.inputs
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.ToBeFixedForConfigurationCache
+import spock.lang.Issue
+
+class InputTrackingIntegrationTest extends AbstractIntegrationSpec {
+
+    @Issue("https://github.com/gradle/gradle/issues/32564")
+    @ToBeFixedForConfigurationCache(
+        because = "values() are not tracked",
+        iterationMatchers = ".*\\.values\\(\\).*"
+    )
+    def "can use instrumented #value as task input"() {
+        given:
+        buildFile """
+            abstract class PrintEnvTask extends DefaultTask {
+                @Input
+                Object toPrint
+
+                @TaskAction
+                def action() {
+                    println String.valueOf(toPrint)
+                }
+            }
+
+            tasks.register("printEnv", PrintEnvTask) {
+                toPrint = ${value}
+            }
+        """
+
+        when:
+        executer.withEnvironmentVars(SOME_PROPERTY: "FIRST_VALUE")
+        run "-DSOME_PROPERTY=FIRST_VALUE", "printEnv"
+
+        then:
+        outputContains(expectedFirstResult)
+
+        when:
+        executer.withEnvironmentVars(SOME_PROPERTY: "SECOND_VALUE")
+        run "-DSOME_PROPERTY=SECOND_VALUE", "printEnv"
+
+        then:
+        outputContains(expectedSecondResult)
+
+        where:
+        value                          | expectedFirstResult         | expectedSecondResult
+        "System.getenv()"              | "SOME_PROPERTY=FIRST_VALUE" | "SOME_PROPERTY=SECOND_VALUE"
+        "System.getenv().entrySet()"   | "SOME_PROPERTY=FIRST_VALUE" | "SOME_PROPERTY=SECOND_VALUE"
+        "System.getenv().keySet()"     | "SOME_PROPERTY"             | "SOME_PROPERTY"
+        "System.getenv().values()"     | "FIRST_VALUE"               | "SECOND_VALUE"
+        "System.properties"            | "SOME_PROPERTY=FIRST_VALUE" | "SOME_PROPERTY=SECOND_VALUE"
+        "System.properties.entrySet()" | "SOME_PROPERTY=FIRST_VALUE" | "SOME_PROPERTY=SECOND_VALUE"
+        "System.properties.keySet()"   | "SOME_PROPERTY"             | "SOME_PROPERTY"
+        "System.properties.values()"   | "FIRST_VALUE"               | "SECOND_VALUE"
+    }
+}

--- a/platforms/core-configuration/input-tracking/src/main/java/org/gradle/internal/configuration/inputs/AccessTrackingEnvMap.java
+++ b/platforms/core-configuration/input-tracking/src/main/java/org/gradle/internal/configuration/inputs/AccessTrackingEnvMap.java
@@ -20,6 +20,7 @@ import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.ForwardingMap;
 import org.jspecify.annotations.Nullable;
 
+import java.io.Serializable;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Set;
@@ -28,7 +29,7 @@ import java.util.function.BiConsumer;
 /**
  * A wrapper for the {@link System#getenv()} result that notifies a listener about accesses.
  */
-public class AccessTrackingEnvMap extends ForwardingMap<String, String> {
+public class AccessTrackingEnvMap extends ForwardingMap<String, String> implements Serializable {
     private final Map<String, String> delegate;
     private final BiConsumer<? super String, ? super String> onAccess;
 
@@ -65,7 +66,7 @@ public class AccessTrackingEnvMap extends ForwardingMap<String, String> {
 
     @Override
     public Set<String> keySet() {
-        return new AccessTrackingSet<>(super.keySet(), trackingListener());
+        return new AccessTrackingSet<>(super.keySet(), keyTrackingListener());
     }
 
     @Nullable
@@ -130,7 +131,7 @@ public class AccessTrackingEnvMap extends ForwardingMap<String, String> {
         delegate.forEach(onAccess);
     }
 
-    private AccessTrackingSet.Listener trackingListener() {
+    private AccessTrackingSet.Listener keyTrackingListener() {
         return new AccessTrackingSet.Listener() {
             @Override
             public void onAccess(@Nullable Object o) {
@@ -176,5 +177,13 @@ public class AccessTrackingEnvMap extends ForwardingMap<String, String> {
                 // Environment variables are immutable.
             }
         };
+    }
+
+    private Object writeReplace() {
+        // When we serialize the whole map, it is likely used as a task input.
+        // We don't know how it is going to be used afterward, and we cannot affect the fingerprint during execution.
+        // Let's be pessimistic and consider everything an input. On a bright side, we don't need access tracking for the deserialized thing.
+        reportAggregatingAccess();
+        return delegate;
     }
 }

--- a/platforms/core-configuration/input-tracking/src/main/java/org/gradle/internal/configuration/inputs/AccessTrackingEnvMap.java
+++ b/platforms/core-configuration/input-tracking/src/main/java/org/gradle/internal/configuration/inputs/AccessTrackingEnvMap.java
@@ -44,7 +44,7 @@ public class AccessTrackingEnvMap extends ForwardingMap<String, String> implemen
     }
 
     @Override
-    public String get(@Nullable Object key) {
+    public @Nullable String get(@Nullable Object key) {
         return getAndReport(key);
     }
 

--- a/platforms/core-configuration/input-tracking/src/main/java/org/gradle/internal/configuration/inputs/AccessTrackingProperties.java
+++ b/platforms/core-configuration/input-tracking/src/main/java/org/gradle/internal/configuration/inputs/AccessTrackingProperties.java
@@ -186,7 +186,7 @@ public class AccessTrackingProperties extends Properties {
     }
 
     @Override
-    public Object putIfAbsent(Object key, Object value) {
+    public @Nullable Object putIfAbsent(Object key, Object value) {
         Object oldValue;
         synchronized (delegate) {
             oldValue = delegate.putIfAbsent(key, value);

--- a/platforms/core-configuration/input-tracking/src/main/java/org/gradle/internal/configuration/inputs/AccessTrackingProperties.java
+++ b/platforms/core-configuration/input-tracking/src/main/java/org/gradle/internal/configuration/inputs/AccessTrackingProperties.java
@@ -96,7 +96,6 @@ public class AccessTrackingProperties extends Properties {
         void onClear();
     }
 
-    // TODO(https://github.com/gradle/configuration-cache/issues/337) Only a limited subset of method is tracked currently.
     private final Properties delegate;
     private final Listener listener;
 
@@ -113,7 +112,7 @@ public class AccessTrackingProperties extends Properties {
 
     @Override
     public Set<String> stringPropertyNames() {
-        return new AccessTrackingSet<>(delegate.stringPropertyNames(), trackingListener());
+        return new AccessTrackingSet<>(delegate.stringPropertyNames(), keyTrackingListener());
     }
 
     @Override
@@ -142,7 +141,7 @@ public class AccessTrackingProperties extends Properties {
 
     @Override
     public Set<Object> keySet() {
-        return new AccessTrackingSet<>(delegate.keySet(), trackingListener());
+        return new AccessTrackingSet<>(delegate.keySet(), keyTrackingListener());
     }
 
     @Override
@@ -556,7 +555,7 @@ public class AccessTrackingProperties extends Properties {
         return clazz == String.class || Primitives.isWrapperType(clazz);
     }
 
-    private AccessTrackingSet.Listener trackingListener() {
+    private AccessTrackingSet.Listener keyTrackingListener() {
         return new AccessTrackingSet.Listener() {
             @Override
             public void onAccess(@Nullable Object o) {

--- a/platforms/core-configuration/input-tracking/src/main/java/org/gradle/internal/configuration/inputs/AccessTrackingSet.java
+++ b/platforms/core-configuration/input-tracking/src/main/java/org/gradle/internal/configuration/inputs/AccessTrackingSet.java
@@ -21,8 +21,13 @@ import com.google.common.collect.Iterators;
 import org.jspecify.annotations.NonNull;
 import org.jspecify.annotations.Nullable;
 
+import java.io.Serializable;
+import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Objects;
 import java.util.Set;
 import java.util.function.Function;
 
@@ -31,7 +36,7 @@ import java.util.function.Function;
  *
  * @param <E> the type of elements
  */
-class AccessTrackingSet<E> extends ForwardingSet<E> {
+class AccessTrackingSet<E> extends ForwardingSet<E> implements Serializable {
     public interface Listener {
         void onAccess(@Nullable Object o);
 
@@ -42,7 +47,6 @@ class AccessTrackingSet<E> extends ForwardingSet<E> {
         void onClear();
     }
 
-    // TODO(https://github.com/gradle/configuration-cache/issues/337) Only a limited subset of entrySet/keySet methods are currently tracked.
     private final Set<? extends E> delegate;
     private final Listener listener;
     private final Function<E, E> factory;
@@ -159,5 +163,31 @@ class AccessTrackingSet<E> extends ForwardingSet<E> {
 
     private void reportAggregatingAccess() {
         listener.onAggregatingAccess();
+    }
+
+    private Object writeReplace() {
+        // When we serialize the whole set, it is likely used as a task input.
+        // We don't know how it is going to be used afterward, and we cannot affect the fingerprint during execution.
+        // Let's be pessimistic and consider everything an input.
+        reportAggregatingAccess();
+        // We cannot just return a LinkedHashSet here, because of a CC bug.
+        // When handling the writeReplace result, CC always serializes it as a Bean, not adhering to any custom protocol.
+        // The LinkedHashSet cannot be serialized correctly as a bean, so we have to wrap it.
+        return new SerializedForm(new ArrayList<>(delegate));
+    }
+
+    /**
+     * Workaround for <a href="https://github.com/gradle/gradle/issues/26942">CC's inability to serialize result of writeReplace properly</a>.
+     */
+    private static class SerializedForm implements Serializable {
+        private final List<?> delegate;
+
+        private SerializedForm(List<?> delegate) {
+            this.delegate = delegate;
+        }
+
+        private Object readResolve() {
+            return new LinkedHashSet<>(Objects.requireNonNull(delegate));
+        }
     }
 }

--- a/platforms/core-runtime/base-services/src/main/java/org/gradle/internal/jvm/JpmsConfiguration.java
+++ b/platforms/core-runtime/base-services/src/main/java/org/gradle/internal/jvm/JpmsConfiguration.java
@@ -83,6 +83,7 @@ public class JpmsConfiguration {
             "--add-opens=java.prefs/java.util.prefs=ALL-UNNAMED", // required by JavaObjectSerializationCodec.kt
             "--add-opens=java.base/java.nio.charset=ALL-UNNAMED", // required by BeanSchemaKt
             "--add-opens=java.base/java.net=ALL-UNNAMED", // required by JavaObjectSerializationCodec
+            "--add-opens=java.base/java.util.concurrent=ALL-UNNAMED", // required by AccessTrackingProperties
             "--add-opens=java.base/java.util.concurrent.atomic=ALL-UNNAMED", // serialized from org.gradle.internal.file.StatStatistics$Collector
             "--add-opens=java.xml/javax.xml.namespace=ALL-UNNAMED" // serialized from IvyDescriptorFileGenerator.Model
         ));

